openedx/core/djangoapps/programs/utils.py

```diff
@@ -460,57 +461,86 @@ def _attach_course_run_upgrade_url(self, run_mode):
 
-        return is_paid_seat
+        Arguments:
+            courses (list): Containing dicts representing courses in a program
+
+        Returns:
+            A subset of the given list of course dicts
+        """
+        enrollments = self.user.courseenrollment_set.filter(
+            is_active=True,
+            mode__in=self.data['applicable_seat_types']
+        )
+        course_runs_with_enrollments = set(unicode(enrollment.course_id) for enrollment in enrollments)
+        courses_without_enrollments = []
+        for course in courses:
+            if len([run for run in course['course_runs'] if unicode(run['key']) in course_runs_with_enrollments]) == 0:
+                courses_without_enrollments.append(course)
+
+        return courses_without_enrollments
 
     def _collect_one_click_purchase_eligibility_data(self):
         """
         Extend the program data with data about learner's eligibility for one click purchase,
         discount data of the program and SKUs of seats that should be added to basket.
         """
-        applicable_seat_types = self.data['applicable_seat_types']
+        applicable_seat_types = set(seat for seat in self.data['applicable_seat_types'] if seat != 'credit')
         is_learner_eligible_for_one_click_purchase = self.data['is_program_eligible_for_one_click_purchase']
         skus = []
         bundle_variant = 'full'
-        if is_learner_eligible_for_one_click_purchase:
-            for course in self.data['courses']:
-                add_course_sku = True
-                course_runs = course.get('course_runs', [])
-                published_course_runs = filter(lambda run: run['status'] == 'published', course_runs)
 
-                if len(published_course_runs) == 1:
-                    for course_run in course_runs:
-                        is_paid_seat = self._check_enrollment_for_user(course_run)
-
-                        if is_paid_seat:
-                            add_course_sku = False
-                            break
-
-                    if add_course_sku:
+        if is_learner_eligible_for_one_click_purchase:
+            courses = self._filter_out_courses_with_enrollments(self.data['courses'])
+            courses = self._filter_out_courses_with_entitlements(courses)
+
+            if len(courses) < len(self.data['courses']):
+                bundle_variant = 'partial'
+
+            for course in courses:
+                entitlement_product = False
+                for entitlement in course.get('entitlements', []):
+                    if entitlement['mode'] in applicable_seat_types:
+                        skus.append(entitlement['sku'])
+                        entitlement_product = True
+                        break
+                if not entitlement_product:
+                    course_runs = course.get('course_runs', [])
+                    published_course_runs = [run for run in course_runs if run['status'] == 'published']
```

Comment:
Do we need to consider encoding when when comparing `run['status']` to `'published'`? Also, is there a constant we can use instead of the string, 'published'?

Reply:
There isn't already a constant, but I can add one.

Reply:
Probably somewhere in the common directory. Maybe see if there's a constants file defined anywhere? It might be worth asking someone with more python/django experience for a recommendation.